packages/glob/src/internal-pattern-helper.ts

```diff
@@ -0,0 +1,142 @@
+import * as pathHelper from './internal-path-helper'
+import {IGlobOptions} from './internal-glob-options'
+import {MatchResult} from './internal-match-result'
+import {Pattern} from './internal-pattern'
+
+const IS_WINDOWS = process.platform === 'win32'
+
+/**
+ * Given an array of patterns, returns an array of paths to search.
+ * Duplicates and paths under other included paths are filtered out.
+ */
+export function getSearchPaths(patterns: Pattern[]): string[] {
+  // Ignore comment and negate patterns
+  patterns = patterns.filter(x => !x.comment && !x.negate)
+
+  // Create a map of all search paths
+  const searchPathMap: {[key: string]: string} = {}
+  for (const pattern of patterns) {
+    const key = IS_WINDOWS
+      ? pattern.searchPath.toUpperCase()
+      : pattern.searchPath
+    searchPathMap[key] = 'candidate'
+  }
+
+  const result: string[] = []
+
+  for (const pattern of patterns) {
+    // Check if already included
+    const key = IS_WINDOWS
+      ? pattern.searchPath.toUpperCase()
+      : pattern.searchPath
+    if (searchPathMap[key] === 'included') {
+      continue
+    }
+
+    // Check for an ancestor search path
+    let foundAncestor = false
+    let tempKey = key
+    let parent = pathHelper.dirname(tempKey)
+    while (parent !== tempKey) {
+      if (searchPathMap[parent]) {
+        foundAncestor = true
+        break
+      }
+
+      tempKey = parent
+      parent = pathHelper.dirname(tempKey)
+    }
+
+    // Include the search pattern in the result
+    if (!foundAncestor) {
+      result.push(pattern.searchPath)
+      searchPathMap[key] = 'included'
+    }
+  }
+
+  return result
+}
+
+/**
+ * Matches the patterns against the path
+ */
+export function match(patterns: Pattern[], itemPath: string): MatchResult {
```

Comment:
internally supports multiple patterns and negate patterns

the public interface only supports passing a single pattern, but basically will be a one-liner to fix that when needed